Introductory_modules/Introduction_to_Ray.ipynb

```diff
@@ -69,26 +68,371 @@
+    "print(\"Object Ref = \" + str(obj_ref))\n",
+    "print(\"Result = \" + str(ray.get(obj_ref)))"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "**Discussion Question**\n",
+    "\n",
+    "Now, the obvious question arises, *why* do we take our ordinary Python function and turn it into a remote function? To see this, think (and experiment below) about if we looped over our functions:\n",
+    "\n",
+    "```python\n",
+    "for i in range(4):\n",
+    "  regular_quick_nap(i)\n",
+    "\n",
+    "for i in range(4):\n",
+    "  remote_quick_nap.remote(i)\n",
+    "```\n",
+    "\n",
+    "- About how long do you expect the regular `for` loop to run (in seconds)?\n",
+    "- Do you expect the remote `for` loop to run faster or slower relative to the regular loop?"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "**Serial vs. Parallel Processes**\n",
+    "\n",
+    "Python scripts by default execute code in a serial manner. This means that if you perform a calculation inside of a `for` loop, each iteration must finish performing the calculation before the next iteration can start. This means that even if your computer has multiple cores, the script can only take advantage of at most a single core.\n",
+    "\n",
+    "By using Ray, we can transform `for` loops into code that takes advantage of all the cores on our machine by giving each core a set of loop iterations to work on. In other words, we can run the iterations of the loop in *parallel*."
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "# Serial Implementation\n",
+    "%%time\n",
+    "results = []\n",
+    "\n",
+    "for i in range(4):\n",
+    "  results.append(regular_quick_nap(i))\n",
+    "\n",
+    "print(results)"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "**Coding Exercise**\n",
+    "\n",
+    "Let's now perform the same computation asynchronously. Notice here that we create a list of object references, append `.remote()` onto the relevant remote task, append each call to our list of references, call `.get` on the object references outside of the loop, and see a runtime improvement."
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "# Parallel Implementation\n",
+    "start_time = time.time()\n",
+    "\n",
+    "obj_refs = []\n",
+    "\n",
+    "for i in range(4):\n",
+    "  obj_refs.append(remote_quick_nap.remote(i))\n",
+    "parallel_results = ray.get(obj_refs)\n",
+    "\n",
+    "print(parallel_results)\n",
+    "\n",
+    "end_time = time.time()\n",
+    "duration = end_time - start_time\n",
+    "\n",
+    "print(duration)"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "### 2. Remote Objects Example\n",
+    "\n",
```

Comment:
May be couple of sentence what object store is in Ray. for example,

"Objects are immutable and can be accessed from anywhere on the cluster, as they are stored in the shared memory. An object ref is essentially a pointer or a unique ID that can refer to a remote object without seeing its value. If you’re familiar with futures in Python, Java, or Scala, Ray object refs are conceptually similar."

Reply:
Done!